proposals/0215-ios-replaykit-streaming.md

```diff
@@ -0,0 +1,95 @@
+    
+    if (@available(iOS 11.0, *) && self.allowReplayKit) {
+        if (![[RPScreenRecorder sharedRecorder] isRecording]) {
+            [[RPScreenRecorder sharedRecorder] startCaptureWithHandler:^(CMSampleBufferRef  _Nonnull sampleBuffer, RPSampleBufferType bufferType, NSError * _Nullable error) {
+                if (error) {
+                    SDLLogD(@"Video stream error %@", error.debugDescription);
+                } else {
+                    [self.streamManager sendVideoData:CMSampleBufferGetImageBuffer(sampleBuffer)];
+                }
+            } completionHandler:^(NSError * _Nullable error) {
+                if (error) {
+                    SDLLogD(@"Video stream error %@", error.debugDescription);
+                }
+            }];
+        }
+    } else {
+        // Fallback on earlier versions
+        UIGraphicsBeginImageContextWithOptions(bounds.size, YES, 1.0f);
+        switch (self.renderingType) {
+            case SDLCarWindowRenderingTypeLayer: {
+                [self.rootViewController.view.layer renderInContext:UIGraphicsGetCurrentContext()];
+            } break;
+            case SDLCarWindowRenderingTypeViewAfterScreenUpdates: {
+                [self.rootViewController.view drawViewHierarchyInRect:bounds afterScreenUpdates:YES];
+            } break;
+            case SDLCarWindowRenderingTypeViewBeforeScreenUpdates: {
+                [self.rootViewController.view drawViewHierarchyInRect:bounds afterScreenUpdates:NO];
+            } break;
+        }
+        
+        UIImage *screenshot = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        
+        CGImageRef imageRef = screenshot.CGImage;
+        CVPixelBufferRef pixelBuffer = [self.class sdl_pixelBufferForImageRef:imageRef usingPool:self.streamManager.pixelBufferPool];
+        if (pixelBuffer != nil) {
+            [self.streamManager sendVideoData:pixelBuffer];
+            CVPixelBufferRelease(pixelBuffer);
+        }
+    };
+    
+}
+```
+
+## Potential downsides
+
+- ReplayKit displays an alert asking the user to allow the app to record the screen.  However, after the user taps this once, it will only ask again on a new app launch after a certain time window
+- The app would need to support horizontal orientation otherwise the screen would be captured in portrait
+- Must be on iOS 11 or greater
+
+## Impact on existing code
+
+* No changes to the Android library
+* The iOS library would need to alter it's code to add a flag to pick between ReplayKit streaming or the current method of streaming (possibly via the ```SDLStreamingMediaConfiguration``` ).  Also it would need to alter SDLCarWindow.m's ```syncFrame``` method that causes the jump in CPU usage (which can be seen via the CPU profiling instrument while running an SDL enabled navigation app by filtering out system methods).  Also the app would need to be forced in horizontal orientation.
```

Comment:
Update to:
 The iOS library would need to alter it's code to add a flag to pick between ReplayKit streaming or the current method of streaming (possibly via the ```SDLStreamingMediaConfiguration``` ).  **Additionally, it** would need to alter SDLCarWindow.m's ```syncFrame``` method that causes the jump in CPU usage (which can be seen via the CPU profiling instrument while running an SDL enabled navigation app by filtering out system methods).  **The app** would need to be forced **in a horizontal** orientation.